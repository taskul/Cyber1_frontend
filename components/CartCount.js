import { CSSTransition, TransitionGroup } from 'react-transition-group';
import styled from 'styled-components';

const Dot = styled.div`
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 2rem;
  margin-left: 3rem;
  background: var(--yellow, #e5d283);
  color: var(--blackBg, #000);
  /* When numbers change they move within the component and using this we set all numbers to have the same width*/
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const AnimationStyles = styled.span`
  /* we are using these classes to create custom style and animation for the CSSTransitions we are using below */
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: transform 0.4s;
    /* when we are rotating an element we want the back face of it to be hidden so we don't see it in animation */
    backface-visibility: hidden;
  }
  /* The reason we have two styling specificaitons enter and exit is because elements switch during re-render and so we style the component that is entering and the one exiting */
  .count-enter {
    /* rotate 0.5turn is half a turn */
    box-shadow: 0 0 35px #a0a14a;
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    /* we put it on top of the new element being rendered. */
    top: 0;
    position: absolute;
    transform: rotateX(0);
    backface-visibility: hidden;
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

export default function CartCount({ count }) {
  return (
    <AnimationStyles>
      <TransitionGroup>
        {/* 
      using CSSTransition for some animations
      also 400 milliseconds for time 
      */}
        <CSSTransition
          unmountOnExit
          classNames="count"
          key={count}
          timeout={{ enter: 400, exit: 400 }}
        >
          <Dot>{count}</Dot>
        </CSSTransition>
      </TransitionGroup>
    </AnimationStyles>
  );
}
