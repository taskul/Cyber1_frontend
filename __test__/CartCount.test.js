// updating props and re-rendering
import { render, screen } from '@testing-library/react';
import wait from 'waait';
import CartCount from '../components/CartCount';

describe('<CartCount />', () => {
  it('Renders', () => {
    // we don't even need expect here because if there any sort of error that's thrown inside of cart count, the test will automatically fail.
    render(<CartCount count={10} />);
  });
  it('Matches snapshot', () => {
    const { container } = render(<CartCount count={11} />);
    expect(container).toMatchSnapshot();
  });
  it('updates via props', async () => {
    const { container, rerender, debug } = render(<CartCount count={11} />);
    expect(container.textContent).toBe('11');
    expect(container).toHaveTextContent('11'); // exactly the same as the last one
    // to update the props and then test the update, we will use rerender
    rerender(<CartCount count="12" />);
    // the way we have it set up with animation is that it renders 11 and then 12 and then 11 disapears creating an animation. The problem is if we run a regular test we get an error because we get '11' back
    // the fix, we need to wait
    await wait(400); // why we put 400 milliseconds in there? because that is what we have set up in  our animation.
    // const twelve = await screen.findByText('12'); // findBy methods look for what is specified multiple times before givng up, which means on the second or third look they may find 12 because at that point 11 will disapear.
    expect(container.textContent).toBe('12');
  });
});
