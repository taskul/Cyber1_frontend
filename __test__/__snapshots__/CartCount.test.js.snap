// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartCount /> Matches snapshot 1`] = `
<div>
  <span
    class="CartCount__AnimationStyles-sc-xxvp4g-1 bVlQYg"
  >
    <div>
      <div
        class="CartCount__Dot-sc-xxvp4g-0 jASszP"
      >
        11
      </div>
    </div>
  </span>
</div>
`;
